sabnzbd/downloader.py

```diff
@@ -861,25 +832,64 @@ def __handle_recv_result(self, nw: NewsWrapper, bytes_received: int = 0, done: b
             server.bad_cons = 0
             server.errormsg = server.warning = ""
 
-            # Update statistics and decode
-            article.nzf.nzo.update_download_stats(sabnzbd.BPSMeter.bps, server.id, nw.data_position)
-            self.decode(article, nw.get_data_buffer(), nw.data_position)
+            # Decode
+            self.decode(article, nw.get_data_buffer())
 
             if sabnzbd.LOG_ALL:
                 logging.debug("Thread %s@%s: %s done", nw.thrdnum, server.host, article.article)
 
             # Reset connection for new activity
             nw.soft_reset()
+
             # Request a new article immediately if possible
             if nw.connected and server.active and not (self.paused or self.shutdown or self.paused_for_postproc):
                 nw.article = server.get_article()
                 if nw.article:
                     self.__request_article(nw)
                     return
-            server.busy_threads.remove(nw)
-            server.idle_threads.append(nw)
+
+            with sabnzbd.decorators.NZBQUEUE_LOCK:
```

Comment:
Same here, we could do with a Downloader-specific lock.

Reply:
Added DOWNLOADER_LOCK for both and added a missing one on `__reset_nw`